pkg/agent/api/delegatedidentity/v1/service.go

```diff
@@ -0,0 +1,228 @@
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// RegisterService registers privileged service on provided server
+func RegisterService(s *grpc.Server, service *Service) {
+	delegatedidentityv1.RegisterDelegatedIdentityServer(s, service)
+}
+
+type Attestor interface {
+	Attest(ctx context.Context) ([]*common.Selector, error)
+}
+
+type Config struct {
+	Log             logrus.FieldLogger
+	Manager         manager.Manager
+	Attestor        workload_attestor.Attestor
+	AuthorizedUsers []string
+}
+
+func New(config Config) *Service {
+	authorizedUsers := map[string]bool{}
+
+	for _, user := range config.AuthorizedUsers {
+		authorizedUsers[user] = true
+	}
+
+	return &Service{
+		manager:         config.Manager,
+		attestor:        endpoints.PeerTrackerAttestor{Attestor: config.Attestor},
+		authorizedUsers: authorizedUsers,
+	}
+}
+
+// Service implements the privileged server
+type Service struct {
+	delegatedidentityv1.UnsafeDelegatedIdentityServer
+
+	manager  manager.Manager
+	attestor Attestor
+
+	// Spiffe IDs of users that are authorized to use this API
+	authorizedUsers map[string]bool
+}
+
+// isCallerAuthorized attests the caller and returns true if its identity is on
+// the authorized users map.
+func (s *Service) isCallerAuthorized(ctx context.Context) (bool, error) {
+	callerSelectors, err := s.attestor.Attest(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	identities := s.manager.MatchingIdentities(callerSelectors)
+
+	for _, identity := range identities {
+		id := identity.Entry.SpiffeId
+
+		if _, ok := s.authorizedUsers[id]; ok {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (s *Service) SubscribeToX509SVIDs(req *delegatedidentityv1.SubscribeToX509SVIDsRequest, stream delegatedidentityv1.DelegatedIdentity_SubscribeToX509SVIDsServer) error {
+	ctx := stream.Context()
+	log := rpccontext.Logger(ctx)
+
+	authorized, err := s.isCallerAuthorized(ctx)
+
+	if err != nil {
+		log.WithError(err).Error("Workload attestation failed")
+		return status.Error(codes.Internal, "workload attestation failed")
+	}
+
+	if !authorized {
+		return status.Error(codes.PermissionDenied, "no authorized")
+	}
+
+	selectors, err := api.SelectorsFromProto(req.Selectors)
+	if err != nil {
+		return err
+	}
+
+	subscriber := s.manager.SubscribeToCacheChanges(selectors)
+	defer subscriber.Finish()
+
+	for {
+		select {
+		case update := <-subscriber.Updates():
+			if err := sendX509SVIDResponse(update, stream, log); err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func sendX509SVIDResponse(update *cache.WorkloadUpdate, stream delegatedidentityv1.DelegatedIdentity_SubscribeToX509SVIDsServer, log logrus.FieldLogger) (err error) {
+	resp, err := composeX509SVIDBySelectors(update)
+	if err != nil {
+		log.WithError(err).Error("Could not serialize X.509 SVID response")
+		return status.Error(codes.Unavailable, "could not serialize response")
```

Comment:
This seems like it should be codes.Internal instead?

Reply:
hmm, this ideia was get from here I think, should it be codes.Internal instead?
https://github.com/spiffe/spire/blob/9af63ce547c1456d882957a2bcffd6e5ae84f429/pkg/agent/endpoints/workload/handler.go#L318-L322

Reply:
Yes I think Unavailable is wrong :(

Some of the code in workload api handler is very old and these things are hard to change compatibility wise. Feels like it makes sense to do the right thing here since delegated identity api is brand new